tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Output.java

```diff
@@ -39,23 +41,30 @@ public int index() {
 
   /** Returns the DataType of the tensor referred to by this Output. */
   @SuppressWarnings("unchecked")
-  public DataType<T> dataType() {
-    return (DataType<T>)operation.dtype(index);
+  public DataType dataType() {
+    return operation.dtype(index);
+  }
+
+  /** Returns the type of the tensor referred to by this Output. */
+  @SuppressWarnings("unchecked")
+  @Override
+  public Class<T> type() {
+    return (Class<T>)TensorTypeRegistry.find(dataType()).typeClass();
   }
 
   /**
    * Returns this Output object with the type {@code Output<U>}. This method is useful when given a
    * value of type {@code Output<?>}.
    *
-   * @param dt any supported tensor data type
+   * @param type any supported tensor type
    * @throws IllegalArgumentException if the actual data type of this object does not match the type
    *     {@code U}.
    */
   @SuppressWarnings("unchecked")
-  public <U extends TType> Output<U> expect(DataType<U> dt) {
-    if (!dt.equals(this.dataType())) {
+  public <U extends TType> Output<U> expect(Class<U> type) {
+    if (type != type()) {
```

Comment:
You mean that the tensor type classes would be loaded multiple times by different class loaders? According to [these guys](https://stackoverflow.com/questions/14257357/loading-a-class-twice-in-jvm-using-different-loaders), unless we do extra-effort to make the classes distinct, the JVM will continue to consider them equal. Do you have some specific use case in mind?

Reply:
Well if they start off in different `URLClassLoader` instances pointed at the same jar file then they will be different, and I think that could happen inside an application container. I don't think we should ever be able to mix them, but given JavaCPP plays tricks with how the libraries are loaded it might happen. Either way, I think it's fine if we decide we don't care about them coming from different class loaders, but it might be worth noting that this check will fail (so if we do get an issue about this bug then there is a note next to it in the code).

Reply:
I don't see what this has to do with JavaCPP. It uses tricks for its own classes to work around issues that happen in containers and OSGi, but it doesn't do anything about user classes. If you're saying that JavaCPP should also take care of cases like this for user classes, then yeah ok, it doesn't do that, but from what I remember, you were against JavaCPP working around limitations of the JDK just because Oracle doesn't approve of those tricks, even though everyone in the industry is using them, right? If you want to blame JavaCPP for something, stop worrying about what everyone does, and please find something else to complain about.

Reply:
It depends if you're doing a findClass in the generated C++ code, as that will pull down something using a class loader but given javacpp loads the same library into the same address space multiple times it's not clear what class loader you'll get back from jni.

However it's not that likely that you're doing that, so it would be difficult to trigger.

And as usual I'm just pointing out that it's library loading behaviour is explicitly against the JNI spec, so what behaviour we get out of the JVM is undefined. It might work on hotspot today, but it's not guaranteed to always work or always have the same behaviour because it's undefined.

Reply:
Ok so I'll classify this as a non-issue for now